web/hooks/use-sort-and-query-params.tsx

```diff
@@ -15,92 +11,52 @@ export type Sort =
   | 'last-updated'
   | 'score'
 
-export function getSavedSort() {
-  // TODO: this obviously doesn't work with SSR, common sense would suggest
-  // that we should save things like this in cookies so the server has them
-  if (typeof window !== 'undefined') {
-    return localStorage.getItem(MARKETS_SORT) as Sort | null
-  } else {
-    return null
-  }
-}
-
-export interface QuerySortOptions {
-  defaultSort?: Sort
-  shouldLoadFromStorage?: boolean
-  /** Use normal react state instead of url query string */
-  disableQueryString?: boolean
-}
-
-export function useQueryAndSortParams({
-  defaultSort = DEFAULT_SORT,
-  shouldLoadFromStorage = true,
-  disableQueryString,
-}: QuerySortOptions = {}) {
-  const router = useRouter()
-
-  const { s: sort, q: query } = router.query as {
-    q?: string
-    s?: Sort
-  }
+type UpdatedQueryParams = { [k: string]: string }
 
-  const setSort = (sort: Sort | undefined) => {
-    router.replace({ query: { ...router.query, s: sort } }, undefined, {
-      shallow: true,
-    })
-    if (shouldLoadFromStorage) {
-      localStorage.setItem(MARKETS_SORT, sort || '')
+function withURLParams(location: Location, params: UpdatedQueryParams) {
+  const newParams = new URLSearchParams(location.search)
+  for (const [k, v] of Object.entries(params)) {
+    if (!v) {
+      newParams.delete(k)
+    } else {
+      newParams.set(k, v)
     }
   }
+  const newUrl = new URL(location.href)
+  newUrl.search = newParams.toString()
+  return newUrl
+}
 
-  const [queryState, setQueryState] = useState(query)
-
-  useEffect(() => {
-    setQueryState(query)
-  }, [query])
+function updateURL(params: UpdatedQueryParams) {
+  // see relevant discussion here https://github.com/vercel/next.js/discussions/18072
+  const url = withURLParams(window.location, params).toString()
+  const updatedState = { ...window.history.state, as: url, url }
+  window.history.replaceState(updatedState, '', url)
+}
 
-  // Debounce router query update.
-  const pushQuery = useMemo(
-    () =>
-      debounce((query: string | undefined) => {
-        const queryObj = { ...router.query, q: query }
-        if (!query) delete queryObj.q
-        router.replace({ query: queryObj }, undefined, {
-          shallow: true,
-        })
-      }, 100),
-    [router]
-  )
+function getStringURLParam(router: NextRouter, k: string) {
+  const v = router.query[k]
+  return typeof v === 'string' ? v : null
+}
 
-  const setQuery = (query: string | undefined) => {
-    setQueryState(query)
-    if (!disableQueryString) {
-      pushQuery(query)
-    }
+export function useQuery(defaultQuery: string, useUrl: boolean) {
+  const router = useRouter()
+  const initialQuery = useUrl ? getStringURLParam(router, 'q') : null
```

Comment:
Seems like this wouldn't work because on the first render the router is usually "not ready" and doesn't have the query properties defined. I could be wrong.

Reply:
That doesn't really make any conceptual sense to me, since obviously the URL is available from time 0 on the client, and the code seems to work fine, but I will double check.

Reply:
FYI, I can verify that when I test, on the first render both on client and server, `router.isReady` is true and `router.query` is populated.